File: scenario/passenger.py
import numpy as np
import random

from parameters import *


class Passenger:
    """Class for passengers."""
    def __init__(self, x, y, dx, dy, env, phase=-1, inf=False, tau=TAU, speed=None):
        if speed == None:
            self.desired_speed = np.random.normal(SPEED, scale=SPEED_STD)
        else:
            self.desired_speed = speed

        self.seat_selection_prob = SEAT_SELECTION_PROB
        self.inf_at_prob = random.random()
        self.rand = random.random()

        self.x = x
        self.y = y
        self.dx = dx    
        self.dy = dy
        self.vx = self.desired_speed
        self.vy = 0
        self.prob = 0
        self.tau = tau
        self.inf = int(inf)*2
        self.phase = phase

        self.step = -1
        self.patience = abs(int(np.random.normal(PATIENCE, PATIENCE_STD)))
        
        self.initialized = False
        self.seat = None
        self.adj_occupied = []
        

        self.aisle_width = env.aisle_width
        self.xmax = env.dimensions[0]
        self.ymax = env.dimensions[1]
        self.y0 = self.ymax/2 - self.aisle_width/2
        self.y1 = self.ymax/2 + self.aisle_width/2
        self.min_standing_x = env.min_standing_x        

    def is_inside(self, x0, y0, x1, y1):
        """
        Returns true if the passenger is inside the rectangle 
        represented by the points x0, y0 and x1, y1 (bottom left and top right corners)
        """
        return x1 > self.x > x0 and y1 > self.y > y0

    def phase_update(self, standing_spots, seat_list, m_adj, doors):
        """Updates the passenger's phases based on it's position and performs 
        the required actions for each phase change (destination updates, seat or 
        standing spot selection and updating the availability status of seats).
        There are 9 phases in total:
            -1 for idle (used for the driver)
            0 for outside the vehicle and heading inside
            1, 2, 3 for movements towards a seat
            4 indicates seating
            5 indicates standing
            6, 7, 8 for exiting
            9 the passenger can be removed from the simulation
        """
        self.step += 1
        dx = self.dx
        dy = self.dy
        phase = self.phase
        change = False

        if self.phase == 6:
            change = True
            if self.y0 < self.y < self.y1 and self.x > self.min_standing_x + 0.2:
                nearest_door = self.nearest_door(doors)
                self.dx = nearest_door[0]
                self.phase = 7
            
        elif self.phase == 7:
            if self.y0 < self.y < self.y1:
                nearest_door = self.nearest_door(doors)
                self.dx = nearest_door[0]
            
            if self.is_inside(dx - 0.5, self.y0, dx + 0.5, self.y1):
                self.dy = - 0.5
            
            if self.is_inside(dx - 0.4, 0, dx + 0.5, 0.4):
                self.phase = 8

        elif self.phase == 8:
            if self.x < dx - 0.6 or self.x > dx + 0.6:
                self.phase = 7

            if self.y < 0:
                self.dy = - 10

            if self.y < -2:
                self.phase = 9

        elif self.phase == 5:
            if self.step % self.patience == 0:
                self.seat_selection(seat_list, standing_spots)

        self.to_phase_zero()

        if self.phase == 3:
            if self.is_inside(dx - 0.05, dy - 0.05, dx + 0.05, dy + 0.05):
                self.phase = 4

        elif self.phase == 2:
            #if stuck reset phase
            if not (self.y0 < self.y < self.y1):
                self.seat = None
                self.adj_occupied = []
                self.phase = 0
                return True

            if self.is_inside(dx - 0.6, dy - 2, dx + 0.6, dy + 2):
                self.dx = self.seat.x
                self.dy = self.seat.y
                self.seat.occupied = True
                self.phase = 3

        elif self.phase == 1:
            if self.y0 < self.y < self.y1:
                self.dx = self.seat.x + 0.3 * np.cos(self.seat.rotation * np.pi / 180)
                self.dy = self.ymax / 2
                self.phase = 2
            
        elif self.phase == 0:
            if not self.initialized:
                nearest_door = self.nearest_door(doors)
                self.dx = nearest_door[0]
                self.dy = self.ymax / 2
                self.initialized = True

            if self.is_inside(0, 0.1, self.xmax, self.ymax):
                self.initialized = False
                self.seat_selection(seat_list, standing_spots)

        # Check if anything used by the crowd simulation has changed
        if self.phase != phase or self.dx != dx or self.dy != dy:
            change = True
        
        return change

    def seat_selection(self, seat_list, standing_spots):
        """Chooses a seat or spot for the passenger.
        Passengers don't have knowledge of other passengers' choices.
        The probability of picking each option is depended on it's utility 
        (the utility functions can be found in 
        'seat.py' and 'spot.py').
        """
        if self.seat_selection_prob > self.rand:
            seat = self.selection(seat_list, self.x)
            if seat != -1:
                self.seat = seat
                self.adj_occupied = [s.occupied for s in self.seat.adjacent]
                self.phase = 1
                return

        if self.phase != 5:    
            spot = self.selection(standing_spots, self.x)
            self.dx = spot.x
            self.dy = spot.y
            self.phase = 5

    def selection(self, options, *args):
        """A simple selection function. 
        Returns an option or -1 if there are 
        no options with utility greater than 0
        """
        utilities = []
        pool = []
        for option in options:
            utility = option.utility(*args)
            if utility > 0:
                utilities.append(utility)
                pool.append(option)  
        
        utility_sum = sum(utilities)
        
        if utility_sum != 0:
            prob = np.asarray(utilities)/utility_sum
            c = np.random.choice(np.arange(0, len(utilities)), p=prob)
            return pool[c]
        
        return -1

    def nearest_door(self, doors):
        """Returns the nearest door to the passenger"""
        nearest = 0
        min_d = 10000000

        for i, d in enumerate(doors):
            dist = (d[0] - self.x)**2
            if dist < min_d:
                min_d = dist
                nearest = i
        return doors[nearest]

    def to_phase_zero(self):
        """Resets the passenger who to phase 0 to reselect a seat or standing area."""
        if self.seat == None or self.phase in (-1, 3, 4):
            return

        if self.seat.occupied:
            self.seat = None
            self.adj_occupied = []
            self.phase = 0
        
        elif self.adj_occupied != [s.occupied for s in self.seat.adjacent]:
            self.seat = None
            self.adj_occupied = []
            self.phase = 0

    def remove(self):
        """When used, the passenger will begin to exit the environment"""
        try:
            self.seat.occupied = False
            self.seat = None
        except:
            pass

        self.adj_occupied = []
        self.phase = 6
        self.dy = self.ymax/2

        if self.x < self.min_standing_x + 0.2:
            self.dx = self.min_standing_x + 0.2
        else:
            self.dx =self.x
        

File: cpp/scenario/passenger.cpp
#include "passenger.h"

#include <vector>
#include <random>

#include "parameters.h"
#include "seat.h"
#include "../utils/utils.h"


using std::vector;


int Passenger::newuid = 0;


Passenger::Passenger(double passenger_x,
                     double passenger_y,
                     double passenger_dx,
                     double passenger_dy,
                     int initial_phase,
                     bool initial_infection,
                     double passenger_tau,
                     double speed) {
    desired_speed = (speed <= 0) ? normal_dist_rand(SPEED, SPEED_STD) : speed;

    seat_selection_prob = SEAT_SELECTION_PROB;
    rand = rand_double();

    inf_at_prob = rand_double();
    
    uid = newuid++;

    x = passenger_x;
    y = passenger_y;
    dx = passenger_dx;
    dy = passenger_dy;
    vx = 0;
    vy = 0;

    new_x = x;
    new_y = y;
    new_vx = vx;
    new_vy = vy;

    prob = 0;
    tau = passenger_tau;
    inf = (int)(initial_infection*2);
    phase = initial_phase;
    seat = nullptr;

    step = 1;

    patience = abs((int)normal_dist_rand(PATIENCE, PATIENCE_STD));

    initialized = false;
    stabilized = false;
}

bool Passenger::is_inside(double x0, double y0, double x1, double y1) {
    return (x1 > x) && (x > x0) && (y1 > y) && (y > y0);
}

void Passenger::phase_update(Environment *environment) {
    double aisle_width = environment->aisle_width;
    double xmax = environment->length;
    double ymax = environment->width;
    double Y0 = ymax/2 - aisle_width/2;
    double Y1 = ymax/2 + aisle_width/2;
    double min_standing_x = environment->min_standing_x;
    vector<vector<double>> doors = environment->doors;

    // update movements
    x = new_x;
    y = new_y;
    vx = new_vx;
    vy = new_vy;
    
    step += 1;
    if (phase == -1) {
        stabilized = true;
        return;
    }

    if (phase == 6) {
        if ((Y0 < y) && (y < Y1) && x > min_standing_x + 0.2) {
            vector<double> nearest = nearest_door(doors);
            dx = nearest[0];
            phase = 7;
        }
    
    } else if (phase == 7) {
        if ((Y0 < y) && (y < Y1)) {
            vector<double> nearest = nearest_door(doors);
            dx = nearest[0];
        }

        if (is_inside(dx - 0.5, Y0, dx + 0.5, Y1))
            dy = - 0.5;
        
        if (is_inside(dx - 0.4, 0, dx + 0.5, 0.4))
            phase = 8;
    
    } else if (phase == 8) {
        if (x < dx - 0.6 || x > dx + 0.6)
            phase = 7;

        if (y < 0) {
            environment->door_timeout = 0;
            dy = - 10;
        }

        if (y < -2)
            phase = 9;
    
    } else if (phase == 5) {
        if (step % patience == 0)
            seat_selection(environment);
    }

    to_phase_zero();

    if (phase == 3) {
        if (is_inside(dx - 0.05, dy - 0.05, dx + 0.05, dy + 0.05))
            phase = 4;
    
    } else if (phase == 2) {
        // if stuck reset phase
        if (!((Y0 < y) && (y < Y1))) {
            seat = nullptr;
            adj_occupied.clear();
            phase = 0;
            bool_vals_update();
            return;
        }

        if (is_inside(dx - 0.6, dy - 2, dx + 0.6, dy + 2)) {
            dx = seat->x;
            dy = seat->y;
            seat->occupied = true;
            phase = 3;
        }
    
    } else if (phase == 1) {
        if ((Y0 < y) && (y < Y1)) {
            dx = seat->x + 0.3 * cos(seat->r * M_PI / 180);
            dy = ymax / 2;
            phase = 2;
        }

    } else if (phase == 0) {
        if (!initialized) {
            vector<double> nearest = nearest_door(doors);
            dx = nearest[0];
            dy = ymax / 2;
            initialized = true;
        }

        if (is_inside(0, 0.1, xmax, ymax)) {
            initialized = false;
            seat_selection(environment);
            environment->door_timeout = 0;
        }
    }

    bool_vals_update();
}

void Passenger::bool_vals_update() {
    if (step%100 == 0 && step > 0) {
        if (phase == 4)
            stabilized = true;
        
        if (std::abs(past_y-y) < 0.2 && std::abs(past_x-x) < 0.2 &&
            (std::abs(vx)+std::abs(vy)) < 0.3*desired_speed)
            stabilized = true;

        past_x = x;
        past_y = y;
    }

    entering = phase == 0;
    exiting = phase == 7 || phase == 8;
    standing = phase == 5;
    walkspace = phase == 1 || phase == 2 || phase == 7 || phase == 5;
    seated = phase == 3 || phase == 4 || phase == 6;
}

void Passenger::seat_selection(Environment *environment) {
    if (seat_selection_prob > rand) {
        vector<double> utilities;
        double utility;
        vector<Seat*> pool;

        for (auto &s : environment->seats) {
            utility = s.utility(x);
            if (utility > 0) {
                utilities.push_back(utility);
                pool.push_back(&s);
            }
        }

        double utility_sum = 0;
        for (auto &util : utilities)
            utility_sum += util;

        if (utility_sum > 0) {
            double sum = 0;
            int cursor = -1;

            while (sum <= rand_double()*utility_sum) {
                cursor++;
                sum += utilities[cursor];
            }

            seat = pool[cursor];

            vector<bool> new_adj_occupied;
            for (auto s : seat->adjacent)
                new_adj_occupied.push_back(s->occupied);

            adj_occupied = new_adj_occupied;
            phase = 1;
            return;
        }
    }

    if (phase != 5) {
        vector<double> utilities;
        vector<Spot> pool;
        double utility;

        for (auto s : environment->spots) {
            utility = s.utility(x);
            if (utility > 0) {
                utilities.push_back(utility);
                pool.push_back(s);
            }
        }

        double utility_sum = 0;
        for (auto& util : utilities)
            utility_sum += util;

        if (utility_sum > 0) {
            double sum = 0;
            double cursor = -1;

            while (sum <= rand_double()*utility_sum) {
                cursor++;
                sum += utilities[cursor];
            }

            dx = pool[cursor].x;
            dy = pool[cursor].y;
            phase = 5;
            return;
        }
    }
}

vector<double> Passenger::nearest_door(vector<vector<double>> doors) {
    int nearest = 0;
    int min_d = 10000000;

    for (int i = 0; i != doors.size(); i++) {
        double dist = pow((doors[i][0] - x), 2);
        if (dist < min_d) {
            min_d = dist;
            nearest = i;
        }
    }
    return doors[nearest];
}

void Passenger::to_phase_zero() {
    if (seat == nullptr || (phase == -1 || phase == 3 || phase == 4))
        return;

    if (seat->occupied) {
        seat = nullptr;
        adj_occupied.clear();
        phase = 0;
        return;
    }

    vector<bool> new_adj_occupied;
    for (auto s : seat->adjacent)
        new_adj_occupied.push_back(s->occupied);

    if (adj_occupied != new_adj_occupied) {
        seat = nullptr;
        adj_occupied.clear();
        phase = 0;
        return;
    }
}

void Passenger::remove(Environment *environment) {
    stabilized = false;

    if (seat != nullptr) {
        seat->occupied = false;
        seat = nullptr;
    }

    phase = 6;
    dy = environment->width/2;
    
    if (x < environment->min_standing_x + 0.2)
        dx = environment->min_standing_x + 0.2;
    else
        dx = x;
}

int Passenger::get_uid() {
    return uid;
}
